ngraph/core/reference/include/ngraph/runtime/reference/gather.hpp

```diff
@@ -27,149 +30,386 @@ namespace ngraph
+                                                                indices_outer_start_corner,
+                                                                indices_outer_end_corner,
+                                                                indices_outer_strides,
+                                                                indices_outer_axis_order);
+
+                    // Create an inner CoordinateTransfrom for "out"
+                    size_t out_inner_ndim = out_ndim - axis;
+                    Shape out_inner_shape(out_shape);
+                    out_inner_shape.erase(out_inner_shape.begin(), out_inner_shape.begin() + axis);
+                    Coordinate out_inner_start_corner(out_inner_ndim, 0);
+                    Coordinate out_inner_end_corner(out_inner_shape);
+                    if (indices_ndim > 0)
+                    {
+                        out_inner_end_corner[indices_ndim - 1] = 1;
+                    }
+                    for (size_t i = indices_ndim; i < out_inner_ndim; i++)
+                    {
+                        out_inner_end_corner[i] = 1;
+                    }
+                    Strides out_inner_strides(out_inner_ndim, 1);
+                    AxisVector out_inner_axis_order(out_inner_ndim);
+                    std::iota(out_inner_axis_order.begin(), out_inner_axis_order.end(), 0);
+                    CoordinateTransform out_inner_transform(out_inner_shape,
+                                                            out_inner_start_corner,
+                                                            out_inner_end_corner,
+                                                            out_inner_strides,
+                                                            out_inner_axis_order);
+
+                    auto out_outer_coord_iter = out_outer_transform.begin();
+                    for (const Coordinate& params_outer_coord : params_outer_transform)
+                    {
+                        if (out_outer_coord_iter == out_outer_transform.end())
+                            break;
+                        const T* params_prime =
+                            &params[params_outer_transform.index(params_outer_coord)];
+                        T* out_outer = &out[out_outer_transform.index(*out_outer_coord_iter)];
+
+                        auto out_inner_coord_iter = out_inner_transform.begin();
+                        for (const Coordinate& indices_outer_coord : indices_outer_transform)
+                        {
+                            if (out_inner_coord_iter == out_inner_transform.end())
+                                break;
+                            const U* indices_prime =
+                                &indices[indices_outer_transform.index(indices_outer_coord)];
+                            T* out_prime =
+                                &out_outer[out_inner_transform.index(*out_inner_coord_iter)];
+                            gather_nd<T, U>(params_prime,
```

Comment:
Using gather_nd is a simple solution, but not efficient enough.
Inside gather_nd we recalculate batch_shape, do some validation, calculate indices_offsets. And, it seems, the results of these calculations and validation steps aren't changed. We should somehow precalculate all necessary data and use algorithm from gather_nd. I think it will require some refactoring of the gather_nd operation.

Reply:
`v0::gather` will be removed when all will be finish.